UnitySDK/Assets/ML-Agents/Scripts/Agent.cs

```diff
@@ -371,10 +384,15 @@ void OnDisable()
         /// to another (fighting) brain when it comes into contact with an enemy.
         /// </remarks>
         /// <param name="givenBrain">New brain to subscribe this agent to</param>
-        public void GiveBrain(Brain givenBrain)
+        public void GiveModel(NNModel model, string behaviorName)
         {
-            brain = givenBrain;
-            ResetData();
+            m_BrainFactoryParameters.model = model;
+            m_BrainFactoryParameters.behaviorName = behaviorName;
+            m_Brain = new LearningBrain(
+                m_BrainFactoryParameters.brainParameters,
+                model,
+                m_BrainFactoryParameters.inferenceDevice,
+                behaviorName);
```

Comment:
It feels strange to me that the agent knows about and can construct all of the concrete IBrain implementations.  Seems like an information leak between these two objects.

Reply:
That is an interesting point. Should I move this logic into a Factory or is the problem deeper than that?

Reply:
I think a Factory is sufficient.

Reply:
I think the instantiation of brains should happen external to the Agent class.

Reply:
Maybe this is necessary for now until we get to the end of the refactor.

Reply:
Made factory now but we should discuss this point further.